domain/application/resource/types.go

```diff
@@ -0,0 +1,164 @@
+	// resource-get will retrieve.
+	Resources []Resource
+
+	// RepositoryResources provides the resource info from charm hub
+	// for each of the application's resources. The information from
+	// hub charm hub is current as of the last time the charm hub
+	// was polled. Each entry here corresponds to the same indexed entry
+	// in the Resources field. An entry may be empty if data has not
+	// yet been retrieve from the repository.
+	RepositoryResources []resource.Resource
+
+	// UnitResources reports the currently-in-use version of file type
+	// resources for each unit.
+	UnitResources []UnitResources
+}
+
+// Resource defines a single resource within a Juju model.
+//
+// Each application will have exactly the same resources associated
+// with it as are defined in the charm's metadata, no more, no less.
+// When associated with the application the resource may have additional
+// information associated with it.
+//
+// A resource may be a "placeholder", meaning it is only partially
+// populated before an upload (whether local or from the charm store).
+// In that case the following fields are not set:
+//
+//	Timestamp, SuppliedBy, SuppliedByType
+//
+// For "upload" placeholders, the following additional fields are
+// not set:
+//
+//	Fingerprint, Size
+type Resource struct {
+	resource.Resource
+
+	// ID uniquely identifies a resource within the model.
+	ID resources.ID
+
+	// ApplicationID identifies the application for the resource.
+	ApplicationID application.ID
+
+	// SuppliedBy is the name of who added the resource to the controller.
+	// The name is a username if the resource is uploaded from the cli
+	// by a specific user. If the resource is downloaded from a repository,
+	// the ID of the unit which triggered the download is used.
+	SuppliedBy string
+
+	// SuppliedByType indicates what type of value the SuppliedBy name is:
+	// application, username or unit.
+	SuppliedByType SuppliedByType
+
+	// Timestamp indicates when this resource was added to the model in
+	// the case of applications or when this resource was loaded by a unit.
+	Timestamp time.Time
+}
+
+// SuppliedByType indicates what the SuppliedBy name represents.
+type SuppliedByType string
+
+const (
+	Unknown     SuppliedByType = "unknown"
+	Application SuppliedByType = "application"
+	Unit        SuppliedByType = "unit"
+	User        SuppliedByType = "user"
+)
+
+// UnitResources contains the list of resources used by a unit.
+type UnitResources struct {
+	// ID is the ID of the unit.
+	ID unit.ID
+
+	// Resources are the resource versions currently in use by this unit.
+	Resources []Resource
+
+	// DownloadProgress indicates the number of bytes of the unit's
+	// resources, identified by name, that have been downloaded so far
+	// by the uniter. This only applies to resources that are currently
+	// being downloaded to the unit. All other resources for the unit
+	// will not be found in the map.
+	DownloadProgress map[string]int64
```

Comment:
Speaking offline, we're fine with keeping this for now, considering the frequency of downloading a resource. But I will say, that I'm openly against this. This feels like it should be done differently. Writing information [will prevent other writers from attempting to write](https://www.sqlite.org/wal.html#concurrency), so it would be expected that we will hit the retry transaction logic a lot during the situation. I would consider this acceptable for singular or low number models with low activity, but when we want to scale to up and over a 100 models, this and things like this will be an issue. Dqlite doesn't currently have the concept of raft groups (or [multi-raft](https://tikv.org/deep-dive/scalability/multi-raft/)), which means a model downloading a resource can cause contention in other models. This isn't limited to the way Juju has setup the databases, this is just a consequence of how dqlite is implemented.